### ISSUE TITLE
Default variables not inherited when including multiple instances of a role from another role

### ISSUE THREAD

**Issue by Rylon** [ansible/ansible#14017]
**Issue Type:**

Bug Report

**Ansible Version:**

2.0.0.2

**Environment:**

Ubuntu 14.04

**Summary:**

Default variables don't appear to be inherited properly for optional vars
when including multiple instances of a role via another role's meta dependencies.

This behaviour differs from Ansible 1.9.

**Steps To Reproduce:**

I have a role which includes multiple instances of another role:

```
# ./roles/my_role/meta/main.yml

---
dependencies:
  - role:          user
    user_state:    present
    user_username: mrbash
    user_fullname: Mr Bash Shell
    user_sudoers:  true

  - role:          user
    user_state:    present
    user_username: mrfalse
    user_fullname: Mr False Shell
    user_sudoers:  true
    user_shell:    /bin/false
```

In `./roles/user/defaults/main.yml`, we set `user_shell` to `/bin/bash`, so if this var
is not defined, that default should be used.

**Expected Results:**

The user `mrbash` should have their shell set to Bash. The user `mrfalse` should have their shell set to `/bin/false`.

```
TASK [user : Shell for mrbashshell] ********************************************
ok: [ubuntu_trusty] => (item=/bin/bash) => {
    "item": "/bin/bash", 
    "msg": "/bin/bash"
}
[...]
TASK [user : Shell for mrfalseshell] *******************************************
ok: [ubuntu_trusty] => (item=/bin/false) => {
    "item": "/bin/false", 
    "msg": "/bin/false"
}
```

**Actual Results:**

Both users have their shell set to `/bin/false`:

```
TASK [user : Shell for mrbashshell] ********************************************
ok: [ubuntu_trusty] => (item=/bin/false) => {
    "item": "/bin/false", 
    "msg": "/bin/false"
}
TASK [user : Shell for mrfalseshell] *******************************************
ok: [ubuntu_trusty] => (item=/bin/false) => {
    "item": "/bin/false", 
    "msg": "/bin/false"
}
```

It doesn't seem to matter which order the two instances of the role are
specified, the result is the same - both users get their shell set to `/bin/false`.

**Edge Case**

There seems to be an edge case where if I define multiple instances of a role
directly in a playbook, then the behaviour works according to the 'Expected Results'
section.

For example:

```
- name: Test Playbook
  hosts: all
  sudo: yes

  roles:
    - role:          user
      user_state:    present
      user_username: mrbashshell
      user_fullname: Mr Bash Shell
      user_sudoers:  true

    - role:          user
      user_state:    present
      user_username: mrfalseshell
      user_fullname: Mr False Shell
      user_sudoers:  true
      user_shell:    /bin/false
```

Results in:

```
TASK [user : Shell for mrbashshell] ********************************************
ok: [ubuntu_trusty] => (item=/bin/bash) => {
    "item": "/bin/bash", 
    "msg": "/bin/bash"
}
[...]
TASK [user : Shell for mrfalseshell] *******************************************
ok: [ubuntu_trusty] => (item=/bin/false) => {
    "item": "/bin/false", 
    "msg": "/bin/false"
}
```

So it appears that this issue only affects roles included via other roles in
their meta dependencies, not roles which are included directly in a playbook.

**Comment by jimi-c** [comment]
Hi @Rylon, could you share a full example to reproduce this? At least the `user` role you're using above. Thanks!

**Comment by Rylon** [comment]
@jimi-c hi, sure thing - I've pushed a full example here: https://github.com/Rylon/ansible_role_debugging

**Comment by jjshoe** [comment]
Related to #14296 ?

**Comment by Rylon** [comment]
@jimi-c Hi, I just wanted to check to see if there had been any movement on this issue, perhaps a workaround or something?

Thanks :)

**Comment by sloonz** [comment]
`private_role_vars = yes` seems to be a possible workaround for this issue. Obviously not an option if `my_role` tasks need variables defined in the `user` role.

**Comment by jimi-c** [comment]
Hi @Rylon, I corrected some issues with roles this morning and it appears that this issue has been resolved as well. After running your example, I have the following users created (output from `getent passwd`):

<pre>
mrbashshell:x:1003:1003:Mr Bash Shell:/home/mrbashshell:/bin/bash
someone:x:1004:1004:Mr Someone:/home/someone:/bin/bash
mrfalseshell:x:1005:1005:Mr False Shell:/home/mrfalseshell:/bin/false
mrtrueshell:x:1006:1006:Mr True Shell:/home/mrtrueshell:/bin/true
</pre>

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by Rylon** [comment]
@jimi-c thanks very much! In which version of Ansible will the changes including this fix be released?

**Comment by jimi-c** [comment]
These should be included in 2.0.2.
